File: backend/app/schemas/networks.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.network import NetworkType


class NetworkCreate(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    network_type: NetworkType
    parent_interface_id: str | None = None
    vlan_id: int | None = Field(default=None, ge=1, le=4094)
    subnet: str | None = None
    gateway: str | None = None
    ip_range: str | None = None
    mtu: int | None = Field(default=None, ge=576, le=9216)
    enabled: bool = True


class NetworkUpdate(BaseModel):
    enabled: bool | None = None
    mtu: int | None = Field(default=None, ge=576, le=9216)


class NetworkRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    network_type: NetworkType
    parent_interface_id: str | None
    vlan_id: int | None
    subnet: str | None
    gateway: str | None
    ip_range: str | None
    mtu: int | None
    docker_network_id: str | None
    docker_network_name: str | None
    parent_device: str | None
    enabled: bool
    last_error: str | None
    created_at: datetime
    updated_at: datetime
    docker_exists: bool = False


class NetworkValidationIssue(BaseModel):
    code: str
    message: str
    severity: str = "error"


class NetworkValidationResult(BaseModel):
    valid: bool
    issues: list[NetworkValidationIssue] = Field(default_factory=list)

File: frontend/src/types/networks.ts
import { z } from "zod";

export const NetworkTypeSchema = z.enum([
  "management",
  "bridge",
  "ipvlan-l2",
  "ipvlan-l3",
  "macvlan",
  "untagged-access",
]);

export const NetworkSchema = z.object({
  id: z.string(),
  name: z.string(),
  network_type: NetworkTypeSchema,
  parent_interface_id: z.string().nullable(),
  vlan_id: z.number().nullable(),
  subnet: z.string().nullable(),
  gateway: z.string().nullable(),
  ip_range: z.string().nullable(),
  mtu: z.number().nullable(),
  docker_network_id: z.string().nullable(),
  docker_network_name: z.string().nullable(),
  parent_device: z.string().nullable(),
  enabled: z.boolean(),
  last_error: z.string().nullable(),
  created_at: z.string(),
  updated_at: z.string(),
  docker_exists: z.boolean(),
});

export const NetworkValidationSchema = z.object({
  valid: z.boolean(),
  issues: z.array(
    z.object({
      code: z.string(),
      message: z.string(),
      severity: z.string(),
    }),
  ),
});

export type Network = z.infer<typeof NetworkSchema>;
export type NetworkType = z.infer<typeof NetworkTypeSchema>;
export type NetworkCreatePayload = {
  name: string;
  network_type: NetworkType;
  parent_interface_id?: string | null;
  vlan_id?: number | null;
  subnet?: string | null;
  gateway?: string | null;
  ip_range?: string | null;
  mtu?: number | null;
  enabled?: boolean;
};
